### PR TITLE
CANS-848 - Search "Autocompleter" refactor

### PR DESCRIPTION
**Pull request by wmitchell-cambria**
## JIRA

[CANS-848 - Refactor Autocomplete component to use Downshift](https://osi-cwds.atlassian.net/browse/CANS-848)

## Description

This is a big PR with a lot of context. I recommend reading the story first for background. The short story is that the previous Autocompleter component copied between Intake and CANS contained a lot of tech debt. The component in Snapshot was the source of several hard-to-figure-out bugs. This refactoring story drops our dependency on `react-autocomplete` in favor of the more flexible `Downshift`.

There are some interesting components in here, including an update to our LoadingBoundary component to handle updating with new data. That change in particular requires careful attention when creating new LoadingBoundaries, as anonymous functions created during `render` can cause data to be re-fetched excessively.

Future work in this area will likely involve factoring out PersonSuggestion and InputField so that different apps can use the main logic with custom rendering.

## Tests

- [x] I used TDD to develop the feature
- [x] I have included unit tests
- [ ] I have included new acceptance tests or modified existing ones
- [ ] I have included other tests

I tried to add some acceptance tests for these, but Capybara is pretty bad at keyboard inputs, and the way that it finds elements causes the menu to collapse. :-(

## Types of changes

- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactoring (No behavioral changes)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

*Please pay attention to the LoadingBoundary changes!*

## Checklist:

- [x] My code follows the code style of this project.
- [x] I ran all my tests locally which were green.
- [x] My code adds no new issues to Code Climate
- [x] I ran all the linters for the project.
- [x] My code is in a stable state ready to be deployable, but not necessarily complete.
- [x] I promise on my honor as a CWDS developer to ensure I don't break the build (or bring donuts if I leave things broken), to check linters, use best practices, to leave the code in better shape than I found it, and to have a good day.

### CODE REVIEW

**Comment by wmitchell-cambria** [comment]
Now that I have a label, retest this please.

File: app/javascript/Application/components/Search/ClientSearch/ClientSearch.js [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    if (item.type === 'action') {
+      this.performAction(item)
+      this.setState({ isOpen: true })
+    } else {
+      this.props.onSelect(item)
+    }
+  }
+
+  performAction = action => {
+    if (action.action === 'Show More Results') {
```
Comment (pokharelmanish) [inline app/javascript/Application/components/Search/ClientSearch/ClientSearch.js]: Can we make `Show More Results` constant?

Comment (wmitchell-cambria) [reply app/javascript/Application/components/Search/ClientSearch/ClientSearch.js]: Good call. I think there are probably a couple other little cleanup things like this that I missed as well. I'm going through the PR now, too.

File: app/javascript/Application/components/Search/ClientSearch/QueryInput/QueryInput.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+class QueryInput extends React.PureComponent {
+  onChange = value => {
+    this.props.onChange({ target: { value } })
+  }
+
+  render() {
+    const { onChange, ...props } = this.props
+    return (
+      <Debouncer callback={this.onChange} callbackPropName="onChange">
+        <InputField {...props} />
```
Comment (pokharelmanish) [inline app/javascript/Application/components/Search/ClientSearch/QueryInput/QueryInput.js]: Don't we need an `id`  in the InputField?

File: app/javascript/Application/components/Search/ClientSearch/QueryResults/Results.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  clearItems()
+
+  const hasMoreResults = items.length < totalResults
+  const nextSearchAfter = items.length > 0 && items[items.length - 1].sort
+
+  return isOpen ? (
+    <div>
+      <div className={'suggestion-header-wrapper'} key="header" aria-live="polite">
+        <SuggestionHeader
+          page={1}
```
Comment (pokharelmanish) [inline app/javascript/Application/components/Search/ClientSearch/QueryResults/Results.js]: `1` ??

File: app/javascript/Application/components/Search/ClientSearch/ClientSearch.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+  onSelect = item => {
+    if (item.type === 'action') {
+      this.performAction(item)
+      this.setState({ isOpen: true })
+    } else {
+      this.props.onSelect(item)
+    }
+  }
+
```
Comment (pokharelmanish) [inline app/javascript/Application/components/Search/ClientSearch/ClientSearch.js]: Shorter verion.

```
  onSelect = item => {
    if (item.type === 'action') {
      this.performAction(item)
      this.setState({ isOpen: true })
    }
    this.props.onSelect(item)
  }
```

File: app/javascript/Application/components/Search/ClientSearch/QueryInput/Debouncer.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+const WAIT = 400
+
+class Debouncer extends Component {
+  constructor(props) {
+    super(props)
+    this.debounce = this.debounce.bind(this)
+  }
+
+  debounce(value) {
```
Comment (pokharelmanish) [inline app/javascript/Application/components/Search/ClientSearch/QueryInput/Debouncer.js]: I think we can use arrow function so we don't need to bind in `constructor`

File: app/javascript/Application/components/Search/ClientSearch/ClientSearch.js (hunk 4 above)
Comment (wmitchell-cambria) [reply app/javascript/Application/components/Search/ClientSearch/ClientSearch.js]: We do not want to call `onSelect` if the `if` branch is hit.

File: app/javascript/Application/components/Search/ClientSearch/QueryInput/Debouncer.js (hunk 5 above)
Comment (wmitchell-cambria) [reply app/javascript/Application/components/Search/ClientSearch/QueryInput/Debouncer.js]: Good catch. This was a carry-over from the previous Debouncer component.

File: app/javascript/Application/components/Search/ClientSearch/ClientSearch.js (hunk 4 above)
Comment (pokharelmanish) [reply app/javascript/Application/components/Search/ClientSearch/ClientSearch.js]: oh, I thought we were returning it.

File: app/javascript/Application/components/Search/ClientSearch/ClientSearch.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+      searchAfter: null,
+      query: {
+        searchTerm: '',
+      },
+    }
+  }
+
+  onStateChange = ({ isOpen }) => {
+    if (isOpen !== undefined) {
+      this.setState({ isOpen: isOpen })
```
Comment (ggomez-cambriasolutions) [inline app/javascript/Application/components/Search/ClientSearch/ClientSearch.js]: couldn't this be `this.setState({ isOpen })` instead?

**Comment by wmitchell-cambria** [comment]
D'aww... thanks guys!

**Pull request merged**